harness/determined/common/api/authentication.py

```diff
@@ -212,24 +159,24 @@ def logout(
     master_address = master_address or util.get_default_master_address()
     token_store = TokenStore(master_address)
 
-    session_user, _ = default_load_user_password(requested_user, None, token_store)
+    user, _ = default_load_user_password(requested_user, None, token_store)
     # Don't log out of DEFAULT_DETERMINED_USER when it's not specified and not the active user.
 
-    if session_user is None:
+    if user is None:
         return
 
-    if session_user == token_store.get_active_user():
+    if user == token_store.get_active_user():
```

Comment:
There's something uncomfortable about `login` being "login, don't worry about the token store", `login_with_cache` being "login and handle the token store" and for `logout` to be "logout and handle the token store."

Reply:
I think this is again coming down to definitions of "log in" and "log out". Under the assumption that I am a reasonably-smart engineer and that I'm having a hard time working out what these terms are supposed to mean (I BTW also think that we're kind of mixing up definitions as we talk about code in these comments, but that could just be because I haven't fully figured it out for myself), I think there should be documentation.

I think it makes sense to document both in docs.determined.ai and also here in docstrings for `login`, `login_with_cache`, and `logout`.

Reply:
I promised to be opinionated, so here I go. Where the opinions below change functionality, if you like the change but don't think they should be brought into this PR, let's put it in [the wiki](https://hpe-aiatscale.atlassian.net/wiki/spaces/~63e2904c8978d7a4353cbcfd/pages/1483931772/Auth+refactoring).

`logout` can do any of a few distinct things:
1. unset the `active_user`
2. post `post_Logout` to master for whatever the backend does
3. drop the logged-out user's token from the `TokenStore`

Opinions:
1. Each of these three things should be mentioned in the function's docstring if done inside the function
1. If a user can't be found to log out, that's an exception.
1. `logout` should modify `active_user` if and only if `login*` do.
1. Execution of `post_Logout` should not depend on whether the `TokenStore` has a token for the logged-out user (especially not invisibly)

Reply:
And for (3), looking at the [CLI log in](https://github.com/determined-ai/determined/blob/92a380f39e1099c1c02761266888c990f6874d49/harness/determined/cli/user.py#L66), it makes sense to me for the CLI to decide about `set_active`. And thus I think it should also be responsible for unsetting it at logout.

Reply:
I agree, done

Reply:
sorry, did not realize the other three options were independent items.

(1) sure, I'll extend the docstring
(2) sure, whatever.  It means the calling code needs to handle the exception, since at some level deletion-like operations need to be idempotent.  I don't care where, it can be in the CLI and SDK instead of in the `logout()` function that they call.
(4) post_Logout can only logout if you have a token to logout with so that doesn't make sense to me.

Reply:
Ooh. Nice point about (4). You're totally right.

That sounds to me like a place that `logout` should also throw a 401-like exception.

Reply:
pretty sure that _is_ the case that (2) is talking about, isn't it?

Reply:
Ok, actually in the current form of Logout, it returns None in the case that it didn't have anybody to log out as, which I like a lot better than raising an exception.

So I'm going to say "no" to (2) since the same information is already being surfaced in a less annoying manner.

Note that this wasn't the case when you originally wrote your comment.  This is bringing back to me why I assumed most of the rest of your comment was resolved by addressing (3).